cs/src/Connections/TunnelBase.cs

```diff
@@ -0,0 +1,126 @@
+﻿// <copyright file="TunnelBase.cs" company="Microsoft">
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VsSaaS.TunnelService;
+
+/// <summary>
+/// Base class for tunnel client and host.
+/// </summary>
+public abstract class TunnelBase : IAsyncDisposable
+{
+    private readonly object reconnectLock = new();
+    private readonly CancellationTokenSource disposeCts = new();
+    private Task? reconnectTask;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="TunnelBase"/> class.
+    /// </summary>
+    /// <param name="trace"></param>
+    public TunnelBase(TraceSource trace)
+    {
+        Trace = Requires.NotNull(trace, nameof(trace));
+    }
+
+    /// <summary>
+    /// Trace to write output to console.
+    /// </summary>
+    protected TraceSource Trace { get; }
+
+    /// <summary>
+    /// Tunnel connector.
+    /// </summary>
+    protected ITunnelConnector? connector;
+
+    /// <summary>
+    /// Get tunnel access scope for this tunnel client or host.
+    /// </summary>
+    protected abstract string TunnelAccessScope { get; }
+
+    /// <summary>
+    /// Event handler for refreshing the tunnel access token.
+    /// The tunnel client or host fires this event when it is not able to use the access token it got from the tunnel.
+    /// </summary>
+    public event EventHandler<RefreshingTunnelAccessTokenEventArgs>? RefreshingTunnelAccessToken;
+
+    /// <summary>
+    /// SSH session reconnected event.
+    /// If the session doesn't support reconnects, the client may try to start a brand new SSH session.
+    /// This event is fired after either reconnection succeeds or a new session is created.
+    /// </summary>
+    public event EventHandler? SshSessionReconnected;
+
+    /// <summary>
+    /// Gets the fresh tunnel access token when Relay service returns 401.
+    /// </summary>
+    protected async Task<string?> RefreshTunnelAccessTokenAsync(CancellationToken cancellation)
```

Comment:
There should be a default behavior that uses the `ITunnelManagementClient`'s user credentials to get a fresh access token, if the management client has a user token callback.

Currently `TunnelHostBase` has a management client instance; `TunnelClientBase` does not but maybe it should.